Adds weaviate handler

abhijitpal1247 opened the pull request:
## Description

Please include a summary of the change and the issue it solves. 

Fixes #7537 

## Type of change

- [x] ⚡ New feature (non-breaking change which adds functionality)

## Verification Process

To ensure the changes are working as expected:

 - [x]   Test Location: tests/handler_tests/test_weaviate_handler.py
 - [x]   Verification Steps: run `pytest tests/handler_tests/test_weaviate_handler.py -v`

## Additional Media:

- [x] I have attached a brief loom video or screenshots showcasing the new functionality or change.

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have appropriately commented on my code, especially in complex areas.
- [x] Necessary documentation updates are either made or tracked in issues.
- [x] Relevant unit and integration tests are updated or added.




https://github.com/mindsdb/mindsdb/assets/69110711/e80647f9-d29f-4990-b523-6df79f6453c2

abhijitpal1247 commented:
some issues with existing vector database structure
- weaviate doesn't json field in its schema
ref: https://weaviate.io/developers/weaviate/config-refs/datatypes

ZoranPandovski commented:
recheckcla

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,402 @@
+            )
+
+        self._client = None
+        self.is_connected = False
```
It would be much better with an annotation
self.is_connected: bool = False

abhijitpal1247 replied:
resolved in 2d930990be47f593eff6be05211d6c34e2395614

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,402 @@
+    def connect(self):
+        """Connect to a weaviate database."""
+        log.logger.error("connected!")
+        if self.is_connected is True:
```
If we take into account that a variable can only be bool

if self.is_connected

abhijitpal1247 replied:
resolved in 2d930990be47f593eff6be05211d6c34e2395614

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,402 @@
+    def disconnect(self):
+        """Close the database connection."""
+
+        if self.is_connected is False:
```
As mentioned above if not self.is_connected

abhijitpal1247 replied:
resolved in 2d930990be47f593eff6be05211d6c34e2395614

abhijitpal1247 commented:
Update:
- I have implemented metadata the following way:
For each table, a metadata table will be created to keep track of metadata fields. Every time an insertion is made into the main table, an insertion is made into metadata table and a reference to metadata is created in the main table.
- I have  also implemented vector search
Attaching the videos showcasing the same:

https://github.com/mindsdb/mindsdb/assets/69110711/4cb95a82-cf53-400a-8a7e-af44202ce6ab


https://github.com/mindsdb/mindsdb/assets/69110711/17a0089d-b494-4e23-a0c8-7f86b4ae1190

ZoranPandovski commented:
recheckcla

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,617 @@
+
+        if (
+            self._client_config["weaviate_url"] is None
+            or self._client_config["weaviate_api_key"] is None
```
not (self._client_config["weaviate_url"] or self._client_config["weaviate_api_key"])

abhijitpal1247 replied:
resolved in b67cb1454fe8803bc304df7430cab6bcd5d670e7

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,617 @@
+    def _get_client(self) -> weaviate.Client:
+        client_config = self._client_config
+        if client_config is None:
+            raise Exception("Client config is not set!")
```
client_config is dict.
if client_config

abhijitpal1247 replied:
tried to resolve in b67cb1454fe8803bc304df7430cab6bcd5d670e7

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,617 @@
+    def check_connection(self):
+        """Check the connection to the Weaviate database."""
+        response_code = StatusResponse(False)
+        need_to_close = not self.is_connected
```
rm additional variable

abhijitpal1247 replied:
resolved in b67cb1454fe8803bc304df7430cab6bcd5d670e7

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,617 @@
+        # we combine all conditions into a single dict
+        all_conditions = (
+            {"operator": "And", "operands": weaviate_conditions}
+            if len(weaviate_conditions) > 1
```
if weaviate_conditions

abhijitpal1247 replied:
resolved in b67cb1454fe8803bc304df7430cab6bcd5d670e7

abhijitpal1247 replied:
Here, if there are more than one conditions, then we are converting them into one single condition

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,617 @@
+        else:
+            vector_filter = None
+
+        for col in fixed_columns:
```
for col in ("id", "embeddings", "distance", "metadata")

abhijitpal1247 replied:
resolved in b67cb1454fe8803bc304df7430cab6bcd5d670e7

pavloshushkov commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,617 @@
+            ]
+        )
+
+        if len(vector_filter) > 0:
```
if vector_filter

abhijitpal1247 commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,617 @@
+                float: "valueIntList",
+                bool: "valueBooleanList",
+            }
+            value_type = value_list_types.get(type(value[0]))
```
tried to resolve in b67cb1454fe8803bc304df7430cab6bcd5d670e7

abhijitpal1247 replied:
I am not sure how to assign a datatype to an empty list or make a query with it. So, I have put a value error over there. Let me know if any other change is required for it.

dusvyat commented on tests/handler_tests/test_weaviate_handler.py:
```diff
@@ -0,0 +1,496 @@
+            PARAMETERS = {{
+                "weaviate_url" : "{"https://sample-q4x5xe6o.weaviate.network"}",
+                "weaviate_api_key": "{"QBHA2JJ3PxXF6Rs8cl3a6IsYjp0jDFh0sehv"}"
+            }}
```
@abhijitpal1247 I would advise against hard coding api keys in tests.

I would recommend we replace this with connection to locally running weaviate docker for testing

abhijitpal1247 replied:
resolved in e8944f0fe38ecffba0a915419bd098a57c562e55

dusvyat commented on mindsdb/integrations/handlers/weaviate_handler/weaviate_handler.py:
```diff
@@ -0,0 +1,633 @@
+
+connection_args_example = OrderedDict(
+    weaviate_url="https://sample-s3x8blpm.weaviate.network",
+    weaviate_api_key="mdMdLfFSbwcWqLI6NUNUseIoIS8yxnY84U5C",
```
let's not hardcode api key please

abhijitpal1247 replied:
resolved in e8944f0fe38ecffba0a915419bd098a57c562e55

dusvyat reviewed (changes requested):
please can we adjust these comments, otherwise looks good

pavloshushkov reviewed (approved):
LGTM 
@dusvyat @ZoranPandovski

dusvyat commented:
@abhijitpal1247 unit tests were passing yday but not seem to be failing. Please check:

```python
ERROR tests/handler_tests/test_weaviate_handler.py::TestWeaviateHandler::test_create_with_select - ImportError: cannot import name 'Last' from 'mindsdb_sql.parser.ast' 
ERROR tests/handler_tests/test_weaviate_handler.py::TestWeaviateHandler::test_insert_into - ImportError: cannot import name 'Last' from 'mindsdb_sql.parser.ast'
ERROR tests/handler_tests/test_weaviate_handler.py::TestWeaviateHandler::test_select_from - ImportError: cannot import name 'Last' from 'mindsdb_sql.parser.ast' 
ERROR tests/handler_tests/test_weaviate_handler.py::TestWeaviateHandler::test_delete - ImportError: cannot import name 'Last' from 'mindsdb_sql.parser.ast' 
```

dusvyat reviewed (approved):
lgtm
